Training PEFT models with new tokens being added to the embedding layers and tokenizer

pacman100 opened the pull request:
### What does this PR do?
1. Training PEFT models with new tokens being added to the embedding layers and tokenizer.
2. Now, users can train loras targeting the embedding layers with the added new tokens. During saving, the embedding layers also need to be saved as the LoRAs were trained on the specific random initialization of the added tokens. To achieve this, they simply need to pass `save_embedding_layers=True` to the `save_pretrained` method. If not explicitly passed, we try our best to guess if the common embedding module names are present in `target_modules` of the config.
3. Currently, `save_pretrained` doesn't save on main process and as such when working in multi GPU environment, it leads to broken checkpoints as many processes are simultaneously writing to it. This is fixed now with `is_main_process` argument which is inline with that of transformers.
4. Many lines were being duplicated by create_or_update_model_card when checkpoints were saved every `n` steps/epochs. Fixed this so that the created README is clean.
5. End to end example and tests added

Todo:
- [x] Add tests
- [x] Add an end to end example

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_1147). All of your documentation changes will be reflected on that endpoint.

younesbelkada reviewed (approved):
Makes sense, thanks!

BenjaminBossan commented on src/peft/peft_model.py:
```diff
@@ -197,7 +197,10 @@ def save_pretrained(
+                self,
+                state_dict=kwargs.get("state_dict", None),
+                adapter_name=adapter_name,
+                base_model_layers_to_save=kwargs.get("base_model_layers_to_save", None),
```
Instead of doing `kwargs.get`, can we add `base_model_layers_to_save` to the signature of `save_pretrained` and also add a docstring for it (same as for `get_peft_model_state_dict`)?

BenjaminBossan commented on src/peft/utils/save_and_load.py:
```diff
@@ -38,6 +40,8 @@ def get_peft_model_state_dict(model, state_dict=None, adapter_name="default", un
         unwrap_compiled (`bool`, *optional*, defaults to `False`):
             Whether to unwrap the model if torch.compile was used.
+        base_model_layers_to_save (`list(str)`, *optional*, defaults to `None`):
+            The list of base model layers to save. For example, ["embed_tokens", "lm_head"]
```
Let's add a sentence here on what situations the user should pass this argument.

BenjaminBossan commented on src/peft/utils/save_and_load.py:
```diff
@@ -72,6 +76,15 @@ def get_peft_model_state_dict(model, state_dict=None, adapter_name="default", un
+        # save the corresponding base embedding layers when loras are applied to the embedding layers
+        # support from version >= 0.6.2
+        if base_model_layers_to_save is not None:
+            if not isinstance(base_model_layers_to_save, list):
+                raise ValueError(f"{base_model_layers_to_save} should be a list of strings")
```
Let's always do these types of checks as early as possible.

BenjaminBossan commented on src/peft/utils/save_and_load.py:
```diff
@@ -72,6 +76,15 @@ def get_peft_model_state_dict(model, state_dict=None, adapter_name="default", un
+            if not isinstance(base_model_layers_to_save, list):
+                raise ValueError(f"{base_model_layers_to_save} should be a list of strings")
+            for k, v in state_dict.items():
+                if any(f"{module_name}.base_layer" in k for module_name in base_model_layers_to_save):
```
I think this will work for now but could fail once we add support for mixed adapters. When we have a mixed adapter, we can have, for example, a `LoraLayer` whose `base_layer` is a `LoHaLayer` whose `base_layer` is `nn.Linear`. Not 100% sure what the correct solution is, maybe check if the `base_layer` is _not_ a `BaseTunerLayer`?

pacman100 commented:
Hello @BenjaminBossan, I have changed the code quite a bit based on the feedback [here] (https://huggingface.slack.com/archives/C04L3MWLE6B/p1700238526932559?thread_ts=1699626507.649479&cid=C04L3MWLE6B) 

```
Would be great if the base_model_layers_to_save could be automatically handled based on what has change or was trainable
```

As there is only need to save the embedding layers when they are resized, I have changed the code to achieve it. 

Also fixed the below issues:
1. Currently, `save_pretrained`  doesn't save on main process and as such when working in multi GPU environment, it leads to broken checkpoints as many processes are simultaneously writing to it. This is fixed now with `is_main_process` argument which is inline with that of transformers.
2. Many lines were being duplicated by `create_or_update_model_card` when checkpoints were saved e very  n steps/epochs. Fixed this so that the created README is clean.

BenjaminBossan reviewed (commented):
Thanks a lot for reworking the PR to make it more user-friendly, as well as for adding the `is_main_process` argument and for fixing the model card. Also well done adding a notebook to illustrate the new feature.

The logic for determining the embedding layer, even if only a few lines of code, was actually more intricate than I thought and took me some time to understand, but it looks correct to me.

I left a couple of comments. I'd say none of them are hard blockers, but please do take a look. I think my main concern would be the comment about when to automatically toggle the saving of embedding layers.

pacman100 commented on src/peft/utils/save_and_load.py:
```diff
@@ -13,18 +13,34 @@
-def get_peft_model_state_dict(model, state_dict=None, adapter_name="default", unwrap_compiled=False):
+def has_valid_embedding_base_layer(layer):
+    """Check if the layer has an embedding base layer"""
+    return hasattr(layer, "base_layer") and isinstance(layer.base_layer, (torch.nn.Linear, torch.nn.Embedding))
```
`lm_head`, i.e., output embedding layers are `nn.Linear`
![Screenshot 2023-11-29 at 5 01 34 PM](https://github.com/huggingface/peft/assets/13534540/634a0ba4-0ecb-4e39-9a19-27b9c89becc8)

pacman100 commented on src/peft/utils/save_and_load.py:
```diff
@@ -100,6 +118,23 @@ def get_peft_model_state_dict(model, state_dict=None, adapter_name="default", un
+        and any(k in config.target_modules for k in EMBEDDING_LAYER_NAMES)
+        and not is_embedding_layer_resized
+    ):
+        warnings.warn("Setting `is_embedding_layer_resized` to `True` as embedding layers found in `target_modules`")
```
> Additionally, for users who are using prompt-learning, this is not automatically detected, right? Would it make sense to do that too?

Couldn't find a way to do it for prompt learning methods as there is no config related to embeddings being targeted or resized.

pacman100 replied:
> Of course, then it would not work automatically anymore, which was the initial goal.

Yes, and then users can't simply use `trainer.push_to_hub()` or `trainer.save_model()` which adds more burden on end-user. I think the current implementation is fine even though we might save embedding when they weren't resized but targeted. Most usecases will add loras to embedding layers when they resize it.

BenjaminBossan replied:
So what about having the options `True` and `"auto"` (or something similar) which work like `True` and `False` now, with `"auto"` being the default, and if a user explicitly passes `False`, we won't save the embeddings.

pacman100 replied:
Done, added `auto`

pacman100 commented on src/peft/peft_model.py:
```diff
@@ -159,6 +159,8 @@ def save_pretrained(
         save_directory: str,
         safe_serialization: bool = True,
         selected_adapters: Optional[List[str]] = None,
+        is_embedding_layer_resized: bool = False,
```
Done, using `save_embedding_layers `

pacman100 commented on src/peft/peft_model.py:
```diff
@@ -190,48 +198,53 @@ def save_pretrained(
-                )
-            else:
-                torch.save(output_state_dict, os.path.join(output_dir, WEIGHTS_NAME))
+            if is_main_process:
```
Done, thanks!

pacman100 commented on src/peft/utils/save_and_load.py:
```diff
@@ -100,6 +118,23 @@ def get_peft_model_state_dict(model, state_dict=None, adapter_name="default", un
+                # support from version >= 0.6.2
+                embedding_module_name = get_embedding_layer_name(model, layer, config.is_prompt_learning)
+                if embedding_module_name:
+                    to_return.update({k: v for k, v in state_dict.items() if embedding_module_name in k})
```
Thank you, added!

pacman100 commented on tests/test_custom_models.py:
```diff
@@ -750,6 +776,25 @@ def test_non_existing_model_card(self):
+        model = ModelEmbWithEmbeddingUtils()
+        config = LoraConfig(target_modules=["emb", "lin0"], init_lora_weights=False)
+        model = get_peft_model(model, config)
+        print(model)
```
done

pacman100 commented on tests/test_custom_models.py:
```diff
@@ -750,6 +776,25 @@ def test_non_existing_model_card(self):
+        print(model)
+
+        with tempfile.TemporaryDirectory() as tmp_dirname:
+            model.save_pretrained(tmp_dirname, is_embedding_layer_resized=True)
```
Done

pacman100 commented on src/peft/peft_model.py:
```diff
@@ -172,6 +174,12 @@ def save_pretrained(
+            selected_adapters (`list(str)`,  *optional*):
+                A list of adapters to be saved. If `None`, will default to all adapters.
+            is_embedding_layer_resized (`bool`, *optional*):
+                If `True`, save the embedding layers in addition to adapter weights.
```
done

pacman100 commented:
Hello @BenjaminBossan, I've addressed all the comments, Thank you!

BenjaminBossan commented on examples/causal_language_modeling/peft_lora_clm_with_additional_tokens.ipynb:
```diff
@@ -7,7 +7,13 @@
+    "2. Finetuning on a specific language wherein language spoecific tokens are added, e.g., korean tokens being added to vocabulary for finetuning LLM on Korean datasets.\n",
+    "3. Instruction finetuning to return outputs in certain format to enable agent behaviour new tokens such as `<|FUNCTIONS|>`, `<|BROWSE|>`, `<|TEXT2IMAGE|>`, `<|ASR|>`, `<|TTS|>`, `<|GENERATECODE|>`, `<|RAG|>`.\n",
+    "\n",
+    "In such cases, you add the Embedding modules to the LORA `target_modules`. PEFT will take care of saving the embedding layers with the new added tokens along with the adapter weights that were trained on the specific initialization of the embeddings weights of the added tokens."
```
Great explanation.

BenjaminBossan commented on tests/test_custom_models.py:
```diff
@@ -750,6 +777,55 @@ def test_non_existing_model_card(self):
         self.assertGreater(len(model_card), 1000)
 
+    @parameterized.expand(["auto", True, False])
+    def test_targetting_lora_to_embedding_layer(self, save_embedding_layers):
```
```suggestion
    def test_targeting_lora_to_embedding_layer(self, save_embedding_layers):
```

BenjaminBossan commented on tests/test_custom_models.py:
```diff
@@ -750,6 +777,55 @@ def test_non_existing_model_card(self):
+            del state_dict
+
+    @parameterized.expand(["auto", True, False])
+    def test_targetting_lora_to_embedding_layer_non_transformers(self, save_embedding_layers):
```
```suggestion
    def test_targeting_lora_to_embedding_layer_non_transformers(self, save_embedding_layers):
```

BenjaminBossan reviewed (approved):
Looks fantastic, great implementation, examples, docs, and tests.

I found a typo in the test name, apart from that this can be merged.

debraj135 commented:
I have a question about this pr. Is it that before this pr was merged it wasn't possible to both do both lora fine-tuning as well as save the lora adapters for `"embed_tokens", "lm_head"` ?

Why I ask is because i see that when I run the snippet included in this comment [here](https://github.com/huggingface/trl/issues/1287#issuecomment-1918439407), I see a massive difference in the size of the saved adapter files between 0.6.2 and 0.7.0

BenjaminBossan commented:
> I still see this issue of size increase with peft 0.10.0. Is the size increase due to an issue or is it fine? Is it going to be fixed or will the behavior remain with increase in size?

Please check out Sourab's [comment](https://github.com/huggingface/trl/issues/1287#issuecomment-1931339986), does it make clear why you see the size increase and what you can do about it? If not, could you please clarify what's still unclear to you?

amitagh commented:
Thanks Benjamin. I understand that emb layers will be saved by default, but does it mean the size inflation is obvious? Earlier too the modules used to be saved explicitly but it was done with lesser size. For cases where token vocab is not changing they too will now have to leave with bigger size?

BenjaminBossan commented:
> I understand that emb layers (embed_tokens & lm_head) will be saved by default

> For cases where token vocab is not changing they too will now have to leave with bigger size? 

No, the embedding is not saved by default. The default setting is `"auto"`, which means PEFT tries to detect if saving the embedding layer is needed or not. If you find that this automatic detection leads to false positives, let us know and we can try to fix that.

Note that you can always pass `False` to `save_embedding_layers` to prevent this. If you use a library that uses PEFT under the hood but does not expose this parameter, open an issue there and ask them to do so.

> Or is it recommended not to save emb layers if there is no change in token vocab?

Indeed, when the vocab is not changed (no fine-tuning, no additional tokens being added), it does not need to be included in the saved file.

> but does it mean the size inflation is obvious?

Sorry, I don't understand this question, could you please clarify?

amitagh commented:
I am Continuous pretraining with Lora a pretrainined LLM for non-English language without extending the tokenizer as it already has the needed tokens. For non-English lang it is recommended to add emb layers. So in such case since i need emb layers which peft will save as save_emb_layers flag is true it will lead to bigger size adaptor?

BenjaminBossan commented:
> I am Continuous pretraining with Lora a pretrainined LLM for non-English language without extending the tokenizer as it already has the needed tokens. For non-English lang it is recommended to add emb layers.

You have to be more precise here: Are you fully fine-tuning the embedding layer -- in that case, it has to be saved, leading to a much larger file size. Or are you training a LoRA weight on top of the embedding layer -- in that case, only those LoRA weights must be saved, not the whole embedding layer, leading to only a slight increase of the file size (compared to not training the embedding at all).

amitagh commented:
Thank Benjamin. I am doing pretraining with Lora on existing pretrained model Gemma-7B for augmenting non-English corpus with a big corpus. I know Full FT would have been good but cost would have been higher.
Hence doing pretraining with lora.
So yes Lora adaptor will be generated and will be saved separately. But since adaptor size is increasing significantly i guess it will slow down training and finally when i merged the adaptor i hope it doesnt increase the merged model size significantly compared to the input base model.
